src/MultiIndex.zig

```diff
@@ -857,6 +972,122 @@ pub fn exportSnapshot(
     try snapshot.buildSnapshot(writer, index, allocator);
 }
 
+fn startIndexRestore(
+    self: *Self,
+    index_name: []const u8,
+    generation: u64,
+    snapshot_url: []const u8,
+) !void {
+    if (!isValidName(index_name)) {
+        return error.InvalidIndexName;
+    }
+
+    self.lock.lock();
+    defer self.lock.unlock();
+
+    // Get or create the IndexRef entry
+    const entry = try self.indexes.getOrPut(self.allocator, index_name);
+
+    if (!entry.found_existing) {
+        // Allocate new entry
+        entry.key_ptr.* = try self.allocator.dupe(u8, index_name);
+        errdefer self.allocator.free(entry.key_ptr.*);
+
+        const ref = try self.allocator.create(IndexRef);
+        errdefer self.allocator.destroy(ref);
+
+        ref.* = .{
+            .index_dir = try self.dir.makeOpenPath(index_name, .{ .iterate = true }),
+            .redirect = IndexRedirect.init(index_name, generation),
+        };
+        entry.value_ptr.* = ref;
+    }
+
+    const ref = entry.value_ptr.*;
+
+    // Check current state
+    switch (ref.load_state) {
+        .restoring => return error.AlreadyRestoring,
+        .ready => {
+            if (ref.index.has_value and !ref.redirect.deleted) {
+                return error.IndexAlreadyExists;
+            }
+        },
+        .failed => {
+            // Can retry after failure - clean up old context
+            if (ref.load_context) |*ctx| {
+                ctx.deinit(self.allocator);
+            }
+            ref.load_context = null;
+            ref.load_error = null;
+        },
+    }
+
+    // Create load context (stack allocated)
+    const snapshot_url_copy = try self.allocator.dupe(u8, snapshot_url);
+    errdefer self.allocator.free(snapshot_url_copy);
+
+    // Update state
+    ref.load_context = LoadContext{
+        .generation = generation,
+        .snapshot_url = snapshot_url_copy,
+    };
+    ref.load_state = .restoring;
+    ref.redirect.version = generation;
```

Comment:
This is dangerous, maybe a better option would be to handle the whole thing in getOrCreateIndex.